feat: add git worktree view

saoudrizwan opened the pull request:
### Description

Adds a complete Git worktree management UI to Cline, enabling users to run parallel Cline sessions across multiple VS Code windows.

| Worktrees View | New Worktree Modal | Merge Modal |
|----------------|-------------------|-------------|
| <img width="406" height="805" alt="Cursor 2025-12-27 08 50 09" src="https://github.com/user-attachments/assets/47ee7c48-cb60-4352-939c-ad5d6e91470c" /> | <img width="404" height="806" alt="Cursor 2025-12-27 08 50 36" src="https://github.com/user-attachments/assets/767a5a4b-844c-4b15-a230-7749c76cbb6f" />| <img width="405" height="806" alt="Cursor 2025-12-27 08 51 32" src="https://github.com/user-attachments/assets/ab152eb0-1b8a-47a8-a4e2-37693bfed38b" /> |


## The Problem

Cline takes over your VS Code window while working on a task. If you want to keep coding while Cline works, or run multiple Cline tasks in parallel, you're stuck.

## The Solution: Git Worktrees

Git worktrees are a native Git feature that lets you have multiple branches checked out simultaneously in separate folders. All worktrees share the same Git history and `.git` directory, but each has its own working directory with different code checked out.

Key concepts:
- **Main worktree**: Your original repository folder where the `.git` directory lives
- **Linked worktrees**: Additional folders you create, each checked out to a different branch  
- **Shared history**: All worktrees share commits, branches, and Git configuration

Unlike regular branch switching, worktrees let you have multiple branches checked out at the same time in different folders. This means you can have VS Code windows open for different features simultaneously.

## Why Worktrees + Cline?

With worktrees, you can:

1. **Run Cline in parallel** - Have Cline work on multiple tasks simultaneously, each in its own worktree and VS Code window
2. **Keep working while Cline works** - Let Cline handle a task in a separate worktree while you continue coding in your main workspace
3. **Isolate experimental changes** - Test risky changes in a worktree without affecting your main branch
4. **Quick context switching** - Jump between features without stashing or committing incomplete work

## Features

### Worktree Management
- **View all worktrees** for the current repository with status indicators (current, main, locked)
- **Create new worktrees** with auto-suggested branch names and paths (click "Auto-fill with suggested values" to get a random branch name like `worktree/cline-abc12` and a path in Documents)
- **Open worktrees** in current window or new VS Code window (recommended for parallel Cline sessions)
- **Delete linked worktrees** with confirmation (branches are preserved, only the working directory is removed)
- **Checkout branches** in existing worktrees

### Merge Worktrees
When done working in a worktree:
1. Click the merge icon on any linked worktree
2. Review merge details in the confirmation modal
3. Choose whether to delete the worktree after merging
4. Click Merge

If conflicts are detected, Cline shows the conflicting files and offers two options:
- **Ask Cline to Resolve & Merge** - Creates a new Cline task with a prompt asking Cline to resolve conflicts, complete the merge, and clean up the worktree
- **Resolve Manually** - Close the modal and handle it yourself

### `.worktreeinclude`: Automatic File Copying

When you create a new worktree, it starts fresh—no `node_modules`, no build artifacts, no IDE settings. Normally you'd need to run `npm install` or similar setup.

The `.worktreeinclude` file solves this by automatically copying specified files to new worktrees.

**How it works:**
1. Create a `.worktreeinclude` file in your repository root
2. Add glob patterns for files you want copied (using `.gitignore` syntax)
3. When Cline creates a new worktree, files matching **both** `.worktreeinclude` **and** `.gitignore` are copied automatically

Only files matched by both are copied—this prevents accidentally duplicating tracked files.

**Example `.worktreeinclude`:**
```gitignore
# Copy node_modules to avoid npm install
node_modules/

# Copy IDE settings
.vscode/

# Copy build cache
.next/
dist/

# Copy environment files (if gitignored)
.env.local
```

The UI shows a tip if you don't have a `.worktreeinclude` file. If you have a `.gitignore`, you can click "Create from .gitignore" to create one pre-filled, then edit it to keep only patterns you want copied.

## Architecture

**Proto definitions** (`proto/cline/worktree.proto`):
- `WorktreeService` with 10 RPC methods: `listWorktrees`, `createWorktree`, `deleteWorktree`, `switchWorktree`, `getAvailableBranches`, `getWorktreeDefaults`, `getWorktreeIncludeStatus`, `createWorktreeInclude`, `checkoutBranch`, `mergeWorktree`
- Messages for worktree state, branch lists, merge results with conflict detection

**Controller handlers** (`src/core/controller/worktree/`):
- One file per RPC method for clean separation
- Uses `simple-git` library for all Git operations

**Core utilities**:
- `src/utils/git-worktree.ts` - Git worktree operations (list, create, delete, get branches)
- `src/utils/worktree-include.ts` - `.worktreeinclude` file parsing and file copying logic using the `ignore` package for glob matching

**UI** (`webview-ui/src/components/worktrees/WorktreesView.tsx`):
- 758-line React component with create form, worktree list, merge modal
- Polls for updates every 3 seconds with smart diffing to prevent UI flickering
- Accessible via "Worktrees" button in Cline sidebar header

**Documentation** (`docs/features/worktrees.mdx`):
- Complete user guide with getting started steps, best practices, and troubleshooting

### Test Procedure

Tested manually across various scenarios:
- Creating worktrees with new branches and custom paths
- Auto-fill defaults generating unique branch names and paths in Documents
- Opening worktrees in new VS Code windows
- Deleting worktrees and verifying cleanup
- Checkout different branches in worktrees
- Merging worktrees back to main with and without conflicts
- Detecting uncommitted changes in source and target worktrees before merge
- "Ask Cline to Resolve" creating a new task with conflict resolution prompt
- Creating `.worktreeinclude` from `.gitignore` template
- Automatic file copying on worktree creation
- Error handling for non-git repos, invalid paths, locked worktrees
- UI polling without flickering via state diffing

### Type of Change

-   [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
-   [x] ✨ New feature (non-breaking change which adds functionality)
-   [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] ♻️ Refactor Changes
-   [ ] 💅 Cosmetic Changes
-   [ ] 📚 Documentation update
-   [ ] 🏃 Workflow Changes

### Pre-flight Checklist

-   [x] Changes are limited to a single feature, bugfix or chore (split larger changes into separate PRs)
-   [ ] Tests are passing (`npm test`) and code is formatted and linted (`npm run format && npm run lint`)
-   [x] I have created a changeset using `npm run changeset` (required for user-facing changes)
-   [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

### Additional Notes

The merge functionality includes several safety checks:
- Detects uncommitted changes in both source and target worktrees before merging
- Aborts merge and reports conflicts cleanly without leaving repo in conflicted state
- Optionally deletes both worktree and branch after successful merge
- "Ask Cline to Resolve" option delegates conflict resolution to Cline with full context
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Adds a Git worktree management UI to Cline, enabling parallel sessions and worktree operations in VS Code.
> 
>   - **Behavior**:
>     - Adds Git worktree management UI in `WorktreesView.tsx` for creating, deleting, and merging worktrees.
>     - Supports opening worktrees in new or current VS Code windows.
>     - Handles merge conflicts with options to resolve manually or via Cline.
>   - **UI Components**:
>     - `CreateWorktreeModal.tsx` and `DeleteWorktreeModal.tsx` for worktree creation and deletion.
>     - `WorktreesView.tsx` for listing and managing worktrees.
>   - **Backend**:
>     - New proto definitions in `worktree.proto` for worktree operations.
>     - Controller handlers in `src/core/controller/worktree/` for RPC methods like `createWorktree`, `deleteWorktree`, and `mergeWorktree`.
>   - **Utilities**:
>     - `git-worktree.ts` for Git worktree operations.
>     - `worktree-include.ts` for handling `.worktreeinclude` file logic.
>   - **Settings**:
>     - Adds `worktreesEnabled` feature flag in `FeatureFlagsService.ts` and `state-keys.ts`.
>   - **Testing**:
>     - E2E tests in `auth.test.ts`, `chat.test.ts`, and `diff.test.ts` for worktree functionality.
>     - Unit tests in `worktree-include.test.ts` for `.worktreeinclude` logic.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=cline%2Fcline&utm_source=github&utm_medium=referral)<sup> for 774c6bf5703b06cee9684767e72f50fa2443ab1d. You can [customize](https://app.ellipsis.dev/cline/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

saoudrizwan commented:
## JetBrains Implementation Notes

This PR adds one new hostbridge method that will need JetBrains implementation:

### `WorkspaceService.openFolder`

**Proto definition** (`proto/host/workspace.proto`):
```protobuf
rpc openFolder(OpenFolderRequest) returns (OpenFolderResponse);

message OpenFolderRequest {
  string path = 1;       // The path to the folder to open
  bool new_window = 2;   // Whether to open in a new window
}

message OpenFolderResponse {
  bool success = 1;
}
```

**VS Code implementation** (`src/hosts/vscode/hostbridge/workspace/openFolder.ts`):
```typescript
export async function openFolder(request: OpenFolderRequest): Promise<OpenFolderResponse> {
  const uri = vscode.Uri.file(request.path)
  await vscode.commands.executeCommand("vscode.openFolder", uri, { forceNewWindow: request.newWindow })
  return OpenFolderResponse.create({ success: true })
}
```

**JetBrains equivalent**: You'll need to open a project at the given path, either in the current window or a new window based on `new_window`. This is used by the "Switch Worktree" functionality to open worktrees in separate IDE windows.

---

Everything else (git worktree operations, `.worktreeinclude` file handling, merge logic) is in host-agnostic code under `src/core/controller/worktree/` and `src/utils/`, so it should work automatically once `openFolder` is implemented.

saoudrizwan commented:
Addressed the performance issue in ddac2b660. The optimization uses two approaches:

1. **Native `cp -r` for directories** - Patterns like `node_modules/` that match entire directories now use the system's native copy command (10-20x faster than Node.js file-by-file)

2. **Parallel file copying** - Individual files are copied in batches of 100 using `Promise.allSettled` (5-10x faster than sequential)

Also parallelized the directory traversal itself with `Promise.all`.

For a typical project with `node_modules`, this should reduce creation time from ~20s to <2s. The symlink approach was considered but opted for copying since worktrees may need to run `npm install` independently.

Added unit tests for the module as well.

abeatrix reviewed (commented):
-[x] When in a non-git repo:

<img width="624" height="262" alt="image" src="https://github.com/user-attachments/assets/247a8b6f-a36e-44f4-80ad-9bd4e1f7444d" />

-[x] When git repo is opened and there are existing worktrees:

<img width="623" height="374" alt="image" src="https://github.com/user-attachments/assets/f5060431-398c-4e49-9c7d-eaf1362e46dd" />

- [ ] the open in current window or Open in new window buttons don't work for the existing worktree?

<img width="258" height="165" alt="image" src="https://github.com/user-attachments/assets/5c70a383-3792-4f5e-9173-72a038c20623" />

- [x] The `Create New Worktree` window is closed automatically after the worktree is created:

<img width="533" height="418" alt="image" src="https://github.com/user-attachments/assets/7bbc78df-b3a1-44a9-b2ac-d98e23e1403d" />

- [x] the Open in current window or Open in new window buttons work for the new worktree.

To be tested:
- [ ] Multi-repos workspace
- [ ] Read file
- [ ] Edit file

saoudrizwan commented:
## Today's Updates

### Performance: Optimized .worktreeinclude file copying
- Rewrote file copying to use streams instead of loading entire files into memory
- Added comprehensive unit tests for worktree-include functionality

### Multi-root and subfolder workspace detection
- Added detection for multi-root workspaces and when VS Code is opened to a subfolder
- Shows appropriate warnings in WorktreesView explaining why worktrees can't be used in these scenarios
- Updated proto definitions with `isMultiRoot` and `isSubfolder` flags

### Create modal improvements
- Auto-fill suggested branch name and path when modal opens
- Added clear (X) buttons to input fields
- Fixed cursor pointer on close button
- Shows .worktreeinclude warning with "Learn more" link when file doesn't exist

### Quick launch button on home page
- Added "New Worktree Window" button on home screen for fast worktree creation
- Shows current branch and path below the button (clickable to navigate to worktrees view)
- Extracted CreateWorktreeModal into its own reusable component

### Auto-open Cline on worktree launch
- When switching to a worktree via quick launch, Cline sidebar automatically opens
- Works for both "new window" and "current window" modes
- Uses direct globalState access at startup to handle cross-window state timing

### Home screen polish
- Refactored HistoryPreview and HomeHeader components
- Fixed word wrap for long branch names and paths (was truncating)
- General style cleanup

arafatkatze commented on docs/features/worktrees.mdx:
```diff
@@ -0,0 +1,270 @@
+
+## Why Use Worktrees with Cline?
+
+Worktrees solve a common problem: **Cline takes over your VS Code window while working on a task**. With worktrees, you can:
```
@juanpflores @saoudrizwan While the documentation is great for a first start and covers the definition of a worktree well what would be really nice is to have tiny little gifs that show exactly what clicks to make so that users can use new worktrees etc. 


We can have that as a followup

saoudrizwan commented on webview-ui/src/components/worktrees/DeleteWorktreeModal.tsx:
```diff
@@ -0,0 +1,88 @@
+				<p className="text-sm text-[var(--vscode-descriptionForeground)] mt-0 mb-3">Are you sure? This will:</p>
+				<ul className="text-sm text-[var(--vscode-descriptionForeground)] mt-0 mb-3 pl-5 list-disc">
+					<li>
+						Delete the branch <span className="font-semibold text-[var(--vscode-foreground)]">{branchName}</span>
```
Good catch! Fixed in b1777fb9b - updated the delete modal copy to accurately describe that only the worktree directory is deleted, not the branch.

Also added an optional checkbox "Also delete branch {branchName}" (unchecked by default) with a warning about unpushed commits. This gives users the option while keeping the safe default.
